### PR TITLE
Refactor hierarchical facets

### PR DESCRIPTION
**Pull request by EreMaijala**
Refactor hierarchical facets to use hierarchical lists with details/summary tags based on [an article describing the mechanism](https://iamkate.com/code/tree-views)

This moves rendering normal phtml templates and removes a lot of the code that subsequently is no longer required.

### CODE REVIEW

**Comment by EreMaijala** [comment]
If this is a too extensive change for the dev branch, I'd be happy to target dev-10.0. In that case I'd suggest we take the styling and html from this PR and apply the changes to facets.js like in #3041.

**Comment by EreMaijala** [comment]
Unfortunately this doesn't seem to work with Safari, so I'll need to do a bit more work.

**Comment by EreMaijala** [comment]
Okay, the latest version should do it. And it only required a tiny bit of JS.

**Comment by demiankatz** [comment]
I'm going to hold off on reviewing this until #3047 is done, since I imagine its shape may change slightly after that work is merged and reconciled. Please feel free to remind me when the time comes!

**Comment by EreMaijala** [comment]
Meanwhile I rebased and re-targeted dev-10.0.

**Review by demiankatz — CHANGES_REQUESTED** [review]
@EreMaijala, looks like this needs some conflict resolution before it can move forward...

**Comment by EreMaijala** [comment]
@demiankatz Resolved!

**Comment by demiankatz** [comment]
Thanks, @EreMaijala! I'm out of the office today but will take another look at this after @sturkel89 has had a chance to do some comparison testing across the themes.

@sturkel89, one additional thing to try, in addition to the usual hierarchical facet tests: in facets.ini, there is a recommendation setting that turns on the "SideFacets" module to display sidebar facets. After doing the regular tests, please edit this file to change SideFacets to SideFacetsDeferred -- this switches to a different version of the sidebar which loads the facet data independently of the main page. Please repeat the tests with this version to be sure that it also works correctly. Thank you!

**Comment by sturkel89** [comment]
**ORIGINAL COMMENT:**
In testing this branch, I noted an error/bug associated with hierarchical facets that I think needs attention apart from the updates specific to this branch. The error comes up in both the test and dev branches. (NB: I couldn't get dev-10.0 to load, so I compared the test branch with plain old dev.)

The error is that an attempt to sort the full results list by **Date Descending** leads to a red box error. Every other sorting option works fine.

The URL is [http://localhost/vufind_test/Search/Results?type=AllFields&sort=year](http://localhost/vufind_test/Search/Results?type=AllFields&sort=year) and the error is
image

It may help with troubleshooting to report that this error is NOT observed when only one aspect of hierarchical facets is enabled:

```
[Results]
hierarchical_facet_str_mv = Hierarchy

[SpecialFacets]
;hierarchical[] = hierarchical_facet_str_mv
```

When I uncomment the SpecialFacets setting, the error occurs.

**ADDITIONAL NOTE: **
If the results list has any number of filters applied, this error does NOT occur!** Every sorting option works, even with two filters applied (unlike in the other hierarchical filters branch I've been testing, when different page lengths are applied.)

**UPDATE**:
Demian helped me troubleshoot this and other sorting errors I've been seeing when hierarchical facets are enabled. He suspected that it was an issue with sorting when certain problematic records were part of the results set, and I have narrowed it down to the three Overdrive records:

image

I will re-test and run future tests with that facet eliminated.

**Comment by demiankatz** [comment]
@sturkel89, when you see the red-box error, you can switch VuFind into development mode to get more detail, which is often helpful. In the test environment you have, this involves editing /etc/apache2/conf-enabled/vufindtest.conf, uncommenting [the line that references development mode](https://github.com/vufind-org/vufind/blob/dev/config/vufind/httpd-vufind.conf#L39), and restarting Apache with `sudo systemctl restart apache2`. If you then refresh the page, you'll get a full trace of the error (unless turning on development mode eliminates the error, which is occasionally the case if the problem is related to caching, since development mode disables all caches).

If @EreMaijala can't reproduce the problem on his end, this is worth doing to see what's going on in your setup. And in any case, keep this in mind for future reference, since it's often helpful when things go wrong!

**Comment by sturkel89** [comment]
I tried Demian's suggestion of going into development mode. The error is "Exception: ID not set!" with lots of other detail; I'm not sure how much or what to report here.

**Review by sturkel89 — COMMENTED** [review]
Aside from the error I reported above (which appears in both test and dev), I think this branch is working well.

I’ve checked the test branch in all three themes, with Exclude facets on and off, and with Or facets on and off. I also tested it with different page lengths.

As far as I can tell, everything behaves the same in the test branch and the dev branch. 

Thus far, I have not been successful in checking SideFacetsDeferred, but I anticipate getting to that aspect tomorrow.

**Review by sturkel89 — CHANGES_REQUESTED** [review]
Correction! I have discovered a bug/difference between the test branch and dev.

I just tested hierarchical facets on the home screen by adding **hierarchical_facet_str_mv = Hierarchy** to the top of the [HomePage] section of facets.ini. The hierarchical facets on the VuFind home screen do not behave properly on the test branch.

### Test branch
In this branch, I can view the facet names and click the arrows to open lower levels and view the tree. BUT, **when I click on any facet, the URL changes while the browser stays on the home screen.**

This is what happens when I click on the link for **level1a**. I'm still on the home screen!

image

### Dev branch
In this branch, I have all the functionality of viewing facet names, browsing the tree. **When I click any of the facets, I am brought to the appropriate results screen.** 

This is what happens when I click on the link for **level1a**.

image


These results are the same regardless of theme in both branches. 

**Update:**
These results are still found when the overdrive.mrc records are eliminated.

**Comment by sturkel89** [comment]
### SideFacets & SideFacetsDeferred report:

I looked at the test and dev branches with SideFacetsDeferred switched on and off, both with and without Javascript enabled in the browser, in all three themes. My tests were all on Chrome on Windows. 

In all cases, I had the Exclude facets turned on, and overdrive.mrc records excluded since I'd found them to not play well with hierarchical facets.

**SideFacetsDeferred results:**
With Javascript ENABLED, everything worked as expected, identically between test and dev. Everything functioned the same as it did with regular SideFacets with JS enabled. 

With Javascript DISABLED, test and dev were identical in not displaying any filters:
image

**Regular SideFacets resuts:**
With Javascript disabled in the browser, I found a difference. 
- In the test branch, the top levels of the hierarchy are displayed, but could not be opened. Clicking on the hierarchical facet links did not have an effect.
image


- In the dev branch, no levels of the hierarchy are displayed.
image

**Comment by EreMaijala** [comment]
> Correction! I have discovered a bug/difference between the test branch and dev.
> 
> I just tested hierarchical facets on the home screen by adding **hierarchical_facet_str_mv = Hierarchy** to the top of the [HomePage] section of facets.ini. The hierarchical facets on the VuFind home screen do not behave properly on the test branch.

This issue should be fixed now.

**Comment by EreMaijala** [comment]
@sturkel89 Your test results with SideFacets/SideFacetsDeferrer seem to be as expected to me.

**Comment by EreMaijala** [comment]
@sturkel89 @demiankatz Regarding the error with Overdriver records, it's because SolrOverdrive has an odd getRawData method that doesn't match setRawData. This needs to be handled as a separate issue, but it's a good catch nonetheless!

**Comment by demiankatz** [comment]
> @sturkel89 Your test results with SideFacets/SideFacetsDeferrer seem to be as expected to me.

Agreed, this makes sense -- and at least the new version can display slightly more information than the old in some situations. :-)

**Comment by sturkel89** [comment]
> > Correction! I have discovered a bug/difference between the test branch and dev.
> > I just tested hierarchical facets on the home screen by adding **hierarchical_facet_str_mv = Hierarchy** to the top of the [HomePage] section of facets.ini. The hierarchical facets on the VuFind home screen do not behave properly on the test branch.
> 
> This issue should be fixed now.

**I've tested this, and it is fixed!**

Is there anything else I should test on this PR?

**Review by sturkel89 — APPROVED** [review]
> > Correction! I have discovered a bug/difference between the test branch and dev.
> > I just tested hierarchical facets on the home screen by adding **hierarchical_facet_str_mv = Hierarchy** to the top of the [HomePage] section of facets.ini. The hierarchical facets on the VuFind home screen do not behave properly on the test branch.
> 
> This issue should be fixed now.

**I've tested this, and it is fixed!**

Is there anything else on this PR that I should test?

File: module/VuFind/src/VuFind/AjaxHandler/PluginManager.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/AjaxHandler/PluginManager.php:54]: I see that the factory has been deleted, but not the class itself. Is this intentional?

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/HomePageFacetsTest.php
Comment (demiankatz) [inline module/VuFind/tests/integration-tests/src/VuFindTest/Mink/HomePageFacetsTest.php:91]: Given the bug that @sturkel89 found and @EreMaijala fixed, should we click on one of these and assert something about where we land?

**Review by demiankatz — CHANGES_REQUESTED** [review]
I think this is close to being finished, though I want to wait until next week to give it a final review with fresher eyes. In the meantime, though, a couple of small observations/suggestions.

File: module/VuFind/src/VuFind/AjaxHandler/PluginManager.php
Comment (EreMaijala) [reply module/VuFind/src/VuFind/AjaxHandler/PluginManager.php:54]: No, accidental. Now deleted the class as well.

File: module/VuFind/tests/integration-tests/src/VuFindTest/Mink/HomePageFacetsTest.php
Comment (EreMaijala) [reply module/VuFind/tests/integration-tests/src/VuFindTest/Mink/HomePageFacetsTest.php:91]: Good idea. I've expanded the test and added one for normal facets as well while at it.

**Review by demiankatz — APPROVED** [review]
Looks good to me now, and all tests are passing! Thanks, @EreMaijala!

**Pull request merged**
